### ISSUE TITLE
extract and log grad norm for individual layers

### ISSUE THREAD

**Issue by stas00** [bigscience-workshop/Megatron-DeepSpeed#149]
the paper NormFormer: Improved Transformer Pretraining with Extra Normalization https://arxiv.org/abs/2110.09456 suggests that under preLN:
> gradients at earlier layers tend to be larger than gradients at later layers

so we want to verify that this is so in our case before acting on it and potentially integrating NormFormer.

So we need to expand the tensorboard and logs to log grad norm for individual layers, perhaps as in the paper we can log 5 layers: 0, 1, int(n_layers/2), -2, -1

For reference see p6 in the paper (graphs and discussion).

Currently only L2 average of all layer grad norms is calculated and logged (a single number).

------------------

Performance-wise this will require doing some extra calculations but not much - it's really just figuring out how to broadcast this info to TB so that multiple-points can be logged at once.

Additionally, we could activate this tool on demand - e.g. after encountering a spike we could roll back to the last good checkpoint and run a cycle with this debug feature enabled.

**Comment by stas00** [comment]
@jaketae, the other simpler approach is to not touch Deepspeed and develop a generic tool that will work with any framework or model. Please have a look at this tool I have developed for params under/overflow debug:
https://github.com/huggingface/transformers/blob/95bab53868a91b4809bd5281a72b5f326853e31f/src/transformers/debug_utils.py#L28

So we want a very similar thing, except for backward, so instead of using `register_module_forward_hook` we want `register_module_full_backward_hook`: https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_full_backward_hook.html#torch.nn.modules.module.register_module_full_backward_hook

And report min/max and additional norm of each layer. 

And then we can optimize it to optionally only run for a few layers instead of all.

Then this tool can be plugged into Megatron at will when we need the debug info.
